File: 프로그래머스/3/12927. 야근 지수/야근 지수.py
import heapq

def solution(n, works):
    max_work = [-w for w in works]
    heapq.heapify(max_work)
    for i in range(n):
        if not max_work:
            break
        work = heapq.heappop(max_work)
        work = -work
        if work > 1:
            heapq.heappush(max_work, -work + 1)
    answer = 0
    for w in max_work:
        answer += w*w
    return answer

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for (int work : works) pq.offer(work);
        
        long answer = 0;
        while (n > 0 && !pq.isEmpty()) {
            int maxWork = pq.poll();
            maxWork--;
            if (maxWork > 0) pq.offer(maxWork);
            n--;
        }
        
        while (!pq.isEmpty()) {
            int remain = pq.poll();
            answer += remain * remain;
        }
        
        return answer;
    }
}